proof/refine/ARM/Reply_R.thy

```diff
@@ -96,4 +95,487 @@ lemma replyRemoveTCB_st_tcb_at'_sym_ref:
+  and ex_nonz_cap_to'[wp]: "(\<lambda>s. ex_nonz_cap_to' t s)"
+  and valid_irq_handlers'[wp]: valid_irq_handlers'
+  and valid_irq_states'[wp]: valid_irq_states'
+  and irqs_masked'[wp]: irqs_masked'
+  and ct_idle_or_in_cur_domain'[wp]: ct_idle_or_in_cur_domain'
+  and pspace_domain_valid[wp]: pspace_domain_valid
+  and gsUntypedZeroRanges[wp]: "\<lambda>s. P (gsUntypedZeroRanges s)"
+  and ksArchState[wp]: "\<lambda>s. P (ksArchState s)"
+  and gsMaxObjectSize[wp]: "\<lambda>s. P (gsMaxObjectSize s)"
+  and sch_act_not[wp]: "sch_act_not t"
+  (wp: crunch_wps)
+
+crunches replyRemoveTCB
+  for ctes_of[wp]: "\<lambda>s. P (ctes_of s)"
+  and aligned'[wp]: pspace_aligned'
+  and distinct'[wp]: pspace_distinct'
+  and ct_not_inQ[wp]: ct_not_inQ
+  and ct_idle_or_in_cur_domain'[wp]: ct_idle_or_in_cur_domain'
+  and cur_tcb'[wp]: "cur_tcb'"
+  and no_0_obj'[wp]: "no_0_obj'"
+  and it'[wp]: "\<lambda>s. P (ksIdleThread s)"
+  and ct'[wp]: "\<lambda>s. P (ksCurThread s)"
+  and ksCurDomain[wp]: "\<lambda>s. P (ksCurDomain s)"
+  and ksDomSchedule[wp]: "\<lambda>s. P (ksDomSchedule s)"
+  and ksInterruptState[wp]: "\<lambda>s. P (ksInterruptState s)"
+  and ksMachineState[wp]: "\<lambda>s. P (ksMachineState s)"
+  and ksDomScheduleIdx[wp]: "\<lambda>s. P (ksDomScheduleIdx s)"
+  and sch_act_simple[wp]: "sch_act_simple"
+  and valid_pde_mappings'[wp]: "valid_pde_mappings'"
+  and untyped_ranges_zero'[wp]: "untyped_ranges_zero'"
+  and ifunsafe'[wp]: "if_unsafe_then_cap'"
+  and global_refs'[wp]: "valid_global_refs'"
+  and valid_arch'[wp]: "valid_arch_state'"
+  and typ_at'[wp]: "\<lambda>s. P (typ_at' T p s)"
+  and vms'[wp]: "valid_machine_state'"
+  and valid_queues'[wp]: valid_queues'
+  and valid_queues[wp]: valid_queues
+  and valid_release_queue[wp]: valid_release_queue
+  and valid_release_queue'[wp]: valid_release_queue'
+  and cap_to': "ex_nonz_cap_to' t"
+  and pspace_domain_valid[wp]: pspace_domain_valid
+  (wp: crunch_wps hoare_vcg_if_lift
+   simp: pred_tcb_at'_def obj_at'_strip if_distribR if_bool_eq_conj)
+
+(* FIXME RT: move to...? *)
+lemma valid_mdb'_lift:
+  "(\<And>P. f \<lbrace>\<lambda>s. P (ctes_of s)\<rbrace>) \<Longrightarrow> f \<lbrace>valid_mdb'\<rbrace>"
+  unfolding valid_mdb'_def
+  apply simp
+  done
+
+lemma updateReply_valid_objs'_preserved_strong:
+  "\<lbrace>valid_objs' and (\<lambda>s. \<forall>r. valid_reply' r s \<longrightarrow> valid_reply' (upd r) s)\<rbrace>
+   updateReply rptr upd
+   \<lbrace>\<lambda>_. valid_objs'\<rbrace>"
+  unfolding updateReply_def
+  apply wpsimp
+  apply (frule(1) reply_ko_at_valid_objs_valid_reply')
+  apply clarsimp
+  done
+
+lemma updateReply_valid_objs'_preserved:
+  "\<lbrace>valid_objs' and K (\<forall>r s. valid_reply' r s \<longrightarrow> valid_reply' (upd r) s)\<rbrace>
+   updateReply rptr upd
+   \<lbrace>\<lambda>_. valid_objs'\<rbrace>"
+  apply (wpsimp wp: updateReply_valid_objs'_preserved_strong)
+  done
+
+lemma replyUnlink_valid_objs':
```

Comment:
I know that I keep adding this one to `wpsimp`s. Did we wanna put it in the `wp` set?